internal/providers/session/service.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2024 Stacklok, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package session
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/stacklok/minder/internal/crypto"
+	"github.com/stacklok/minder/internal/db"
+	"github.com/stacklok/minder/internal/providers"
+)
+
+type ProviderSessionService interface {
+	CreateProviderFromSessionState(ctx context.Context, providerClass db.ProviderClass, encryptedCreds *crypto.EncryptedData, state string) (*db.Provider, error)
+}
+
+type providerSessionService struct {
+	sessionStore SessionStore
+	provStore    providers.ProviderStore
+	dbStore      db.Store
+}
+
+func NewProviderSessionService(sessions SessionStore, provStore providers.ProviderStore) ProviderSessionService {
+	return &providerSessionService{
+		sessionStore: sessions,
+		provStore:    provStore,
+	}
+}
+
+func (pss *providerSessionService) CreateProviderFromSessionState(ctx context.Context, providerClass db.ProviderClass, encryptedCreds *crypto.EncryptedData, state string) (*db.Provider, error) {
+	stateData, err := pss.sessionStore.GetBySessionState(ctx, state)
+	if err != nil {
+		return nil, fmt.Errorf("error getting state data by session state: %w", err)
+	}
+
+	// FIXME: This method feels like it should be part of the class manager?
```

Comment:
Could we reuse the `CreateFromConfig` method from the provider class manager here?

Reply:
I will push a hopefully nicer version shortly.